### ISSUE TITLE
Inconsistent behaviour with forward + cache

### ISSUE THREAD

**Issue by tugtugtug** [coredns/coredns#4189]
**What happened**:
When forward is configured with cache, the nslookup and dig with dnssec turned off gets the rrsig in the answer the first time. And subsequent queries (from cache) is correct.

**What you expected to happen**:
If the client isn't enabling dnssec, the rrsig should not be in the response.

**How to reproduce it (as minimally and precisely as possible)**:
$ dig @127.0.0.1 test.my.salesforce.com

Without cache,
```
; <<>> DiG 9.11.5-P4-5.1+deb10u2-Debian <<>> @127.0.0.1 test.my.salesforce.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 23065
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;test.my.salesforce.com.		IN	A

;; ANSWER SECTION:
test.my.salesforce.com.	300	IN	CNAME	na132.my.salesforce.com.
na132.my.salesforce.com. 30	IN	CNAME	na132-ia2.my.salesforce.com.
na132-ia2.my.salesforce.com. 300 IN	CNAME	na132-ia2.ia2.r.my.salesforce.com.
na132-ia2.ia2.r.my.salesforce.com. 30 IN A	13.110.34.153
na132-ia2.ia2.r.my.salesforce.com. 30 IN A	13.110.35.25
na132-ia2.ia2.r.my.salesforce.com. 30 IN A	13.110.32.153

;; Query time: 9 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Oct 09 18:47:33 UTC 2020
;; MSG SIZE  rcvd: 395
```
With cache, the first response will have rrsig, and subsequent works:
```
; <<>> DiG 9.11.5-P4-5.1+deb10u2-Debian <<>> @127.0.0.1 test.my.salesforce.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 41191
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 2048
;; QUESTION SECTION:
;test.my.salesforce.com.		IN	A

;; ANSWER SECTION:
test.my.salesforce.com.	30	IN	CNAME	na132.my.salesforce.com.
test.my.salesforce.com.	30	IN	RRSIG	CNAME 13 4 300 20201207130416 20201008124156 6736 my.salesforce.com. m73UH/xkVHLpt3y/oZiMq9g6RqYvuqwMVSBWE+wDP0TXhh0kAj/xEbtH NkCy6xLd1s1qP/gAqxEOfnDEslDHgw==
na132.my.salesforce.com. 30	IN	CNAME	na132-ia2.my.salesforce.com.
na132.my.salesforce.com. 30	IN	RRSIG	CNAME 13 4 30 20201105060241 20200906055748 6736 my.salesforce.com. 0kCMiLBmLkhgOHIjqv7MqrCqDn3T10tDFk0g4KQ9v2KPnLpN2PYALHH7 lDZ1N55yEbSfH4YuZMQCi0u0eGYUjg==
na132-ia2.my.salesforce.com. 30	IN	CNAME	na132-ia2.ia2.r.my.salesforce.com.
na132-ia2.my.salesforce.com. 30	IN	RRSIG	CNAME 13 4 300 20201117114346 20200918112838 6736 my.salesforce.com. cpYUyKEhx+8Mv/gs4f9alIYuPbZ+ekv5ql3XdigPnk/1PlPRRLv+V1Iu X0PnuTwkyWgNsvkF+Abm2DiyaygZvw==

;; Query time: 10 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Oct 09 18:45:55 UTC 2020
;; MSG SIZE  rcvd: 659
```
**Anything else we need to know?**:
Seems reproducible consistently.

**Environment**:

- the version of CoreDNS:
```
CoreDNS-1.7.1
linux/amd64, go1.15.2, aa82ca6
```

- Corefile:
```
(common) {
    bind 10.10.10.1 127.0.0.1 
    acl {
      allow net 127.0.0.1 10.10.10.1/23
      block
    }
    prometheus
    log
    errors
}
#ZONE-START
#ZONE-END
. {
  forward . 8.8.8.8 1.1.1.1 {
    health_check 10s
  }
  cache
  import common
}
```
- logs, if applicable:
without cache:
```
[INFO] 127.0.0.1:58859 - 36887 "A IN test.my.salesforce.com. udp 69 false 4096" NOERROR qr,rd,ra 407 0.003900966s
```
with cache:
```
[INFO] 127.0.0.1:46247 - 32441 "A IN test.my.salesforce.com. udp 69 true 2048" NOERROR qr,rd,ra 683 0.010361962s
```
- OS (e.g: `cat /etc/os-release`):
```
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
- Others:

**Comment by miekg** [comment]
Might be this https://github.com/coredns/coredns/issues/4146
(will release 1.8.0 soonish that will fix this)

**Comment by sfudeus** [comment]
fwiw: I observe the same (?) thing in my home setup. With forward (to cloudflare) and cache, I observe that when asking for A, I receive only RRSIG, seemingly only upon cache miss.

Disabling cache solves the issue. 

```

; <<>> DiG 9.10.6 <<>> oauth2.gmx.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 17297
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 2048
; PAD: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ("..............................................................................................................")
;; QUESTION SECTION:
;oauth2.gmx.net.			IN	A

;; ANSWER SECTION:
oauth2.gmx.net.		30	IN	RRSIG	A 8 3 300 20201019065308 20201005065308 13905 gmx.net. PmAt8IKrRA6gGhdGgqdKasbi6ZNbzaYUXFqGP8DKjFKXC3BFG0llKHvu Jbd6GzeEWVss5V4GI5Kl9WVYLbu3JZ2zMtIGnHi8563Pvz2WHSXpxS8Z /LXheF8a+MaGjlm7cl7cGH6Ms5X/EpgjXlpTKk/6gUhCSrhzKlnl8/Km RlkyqpXs2alLk/klA1TiJPjxuKuvKlKiSUS/zyxbsxvCAITlRvzib23v vBEABDvx64lvmVJKqkKTWAFQ4xVUga6ImPzpbZ0chGH/uUZkNcwNeGgt JnjTr9luuXBBbxvORAaKyyPs2MAOiGmo0rg4N2wzGz1KyfTNHIWT7F91 b9pysQ==

;; Query time: 51 msec
;; SERVER: 192.168.30.52#53(192.168.30.52)
;; WHEN: Sat Oct 10 11:34:33 CEST 2020
;; MSG SIZE  rcvd: 466

...

; <<>> DiG 9.10.6 <<>> oauth2.gmx.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 51646
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;oauth2.gmx.net.			IN	A

;; ANSWER SECTION:
oauth2.gmx.net.		7	IN	A	217.72.199.48

;; Query time: 6 msec
;; SERVER: 192.168.30.52#53(192.168.30.52)
;; WHEN: Sat Oct 10 11:34:56 CEST 2020
;; MSG SIZE  rcvd: 73

```
Relevant log lines:
```
coredns-coredns-9765cdb5-g2pbb coredns [INFO] 10.42.3.1:49751 - 60803 "A IN oauth2.gmx.net. udp 43 true 2048" NOERROR qr,rd,ra,ad 468 0.013620345s
coredns-coredns-9765cdb5-g2pbb coredns [INFO] 10.42.3.1:57615 - 42953 "A IN oauth2.gmx.net. udp 43 false 4096" NOERROR qr,aa,rd,ra 62 0.000251537s

**Comment by miekg** [comment]
should be fixed with 1.8.0

**Comment by uthark** [comment]
@miekg any ETA on 1.8.0 release?

**Comment by sfudeus** [comment]
@uthark already done

**Comment by tugtugtug** [comment]
@miekg thank you for the quick turnaround. So far it looks good in our test environment.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Inconsistent ..." ]
>@miekg thank you for the quick turnaround. So far it looks good in our test
>environment.

nice! :)

this change was (sadly) way more tricky than expected, good to hear it's working well now

/close

**Comment by nerzhul** [comment]
We have EXACTLY same issue on salesforces (only) on coredns 1.7.1, FYI.
We can disable cache or upgrade to 1.8.0, we will do tests

**Comment by cjellick** [comment]
The k3s project has had some users hit this bug and we're trying to figure out the best course of action:
- downgrade to 1.6.9
- upgrade to 1.8.0
- wait for 1.8.1 and upgrade to that

1.7.1 to 1.8.0 has a significant number of commits.
1.8.0 to 1.8.1 has a good bit too and some of those appear to fix regressions in 1.8.0

Any guidance on the best version to shoot for? (We are talking about kubernetes 1.19 here)

**Comment by chrisohaver** [comment]
@cjellick,  My 2 cents...

The k8s version shouldn't matter.
I think 1.8.0 is a good choice, given that it is already released, and I don't see anything _too_ critical queued up in 1.8.1.
There will be a kubernetes related metric fix in 1.8.1 (#4255), which fixes a regression from 1.7.0.  If that is important, then waiting for 1.8.1 is probably best.
